presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/ApplyTableScanRedirection.java

```diff
@@ -144,70 +157,118 @@ public Result apply(TableScanNode scanNode, Captures captures, Context context)
+            PlanNodeIdAllocator idAllocator,
+            Set<Symbol> requiredSymbols,
+            Map<Symbol, Cast> casts,
+            PlanNode source)
+    {
+        if (casts.isEmpty() && requiredSymbols.equals(ImmutableSet.copyOf(source.getOutputSymbols()))) {
+            return source;
         }
 
-        return Result.ofPlanNode(
-                new ProjectNode(
-                        context.getIdAllocator().getNextId(),
-                        filterNode,
-                        Assignments.identity(scanNode.getOutputSymbols())));
+        return new ProjectNode(
+                idAllocator.getNextId(),
+                source,
+                Assignments.builder()
+                        .putIdentities(Sets.difference(requiredSymbols, casts.keySet()))
+                        .putAll(casts)
+                        .build());
     }
 
-    private static void throwTypeMismatchException(
+    private Cast getCast(
+            boolean allowCoercions,
             CatalogSchemaTableName destinationTable,
             String destinationColumn,
             Type destinationType,
+            Symbol destinationSymbol,
             TableHandle sourceTable,
             ColumnHandle sourceColumnHandle,
             Type sourceType)
     {
-        throw new PrestoException(TYPE_MISMATCH, format(
-                "Redirected column %s.%s has type %s, different from source column %s.%s type: %s",
-                destinationTable,
-                destinationColumn,
-                destinationType,
-                // TODO report source table and column name instead of ConnectorTableHandle, ConnectorColumnHandle toString
-                sourceTable,
-                sourceColumnHandle,
-                sourceType));
+        if (!allowCoercions) {
+            throw new PrestoException(TYPE_MISMATCH, format(
+                    "Redirected column %s.%s has type %s, different from source column %s.%s type: %s",
+                    destinationTable,
+                    destinationColumn,
+                    destinationType,
+                    // TODO report source table and column name instead of ConnectorTableHandle, ConnectorColumnHandle toString
+                    sourceTable,
+                    sourceColumnHandle,
+                    sourceType));
+        }
+
+        try {
+            metadata.getCoercion(destinationType, sourceType);
+        }
+        catch (PrestoException e) {
+            throw new PrestoException(FUNCTION_NOT_FOUND, format(
+                    "Cast not possible from redirected column %s.%s with type %s to source column %s.%s with type: %s",
+                    destinationTable,
+                    destinationColumn,
+                    destinationType,
+                    sourceTable,
+                    sourceColumnHandle,
+                    sourceType));
+        }
+
+        return new Cast(destinationSymbol.toSymbolReference(), toSqlType(sourceType));
```

Comment:
This is missing the typeOnlyCoercion flag.

Reply:
I know. But currently it's saying every coercions is "not type-only coercion". It needs to pass the true/false flag depending on whether that's the case.

Reply:
@martint shouldn't this not be a concern of this rule?
ie shouldn't we have a canonicalizer which can determine a certain cast is type-only?
it's a function of types being converted from/to, not a function of particular creating site

Reply:
Well, the notion of type-only casts in the planner has always been somewhat of a misfeature. It's not clear it even does anything (a type-only cast will be just a passthrough method in the implementation, so the JIT should be able to inline and remove them).

But we don't have any such thing today, and we rely on the callsites to construct the cast properly. In a way, you can think of an IR cast as having `(from, to, type-only)` elements (vs the AST cast, which only has (`from`, `to`))

TL;DR, we should consider getting rid of type-only casts, but in the meantime, we should construct them properly if at all possible.

Reply:
> TL;DR, we should consider getting rid of type-only casts,

can you please file an issue for that?

Reply:
https://github.com/trinodb/trino/issues/8111

Reply:
added call to `isTypeOnlyCoercion`